36-vending-machine-monitor/firmware/vending-machine/vending-machine.ino

```diff
@@ -0,0 +1,1125 @@
+            columnStateChanged(prevItems, items);
+        }
+    }
+
+    /**
+     * @brief Adds fields to the body of an event for the state of this column.
+     * 
+     * @param items The state of the dispensing items
+     * @return J* The created event body
+     */
+    J* buildColumnEvent(const DispensingItems& items) {
+        J* body = JCreateObject();
+        JAddNumberToObject(body, DATA_FIELD_COL, columnNumber());
+        JAddStringToObject(body, DATA_FIELD_NAME, columnName());
+        JAddStringToObject(body, DATA_FIELD_STATE, items.fillStateString());
+        if (items.isOnline()) {
+            JAddNumberToObject(body, DATA_FIELD_DISTANCE, this->distance.distance_mm);
+            JAddNumberToObject(body, DATA_FIELD_ITEMS, items.itemCount);
+        }
+        return body;
+    }
+
+    /**
+     * @brief Creates and sends and event about the current column state, and what changed
+     *  since the last reported event.
+     * 
+     * @param prev      The previous column state
+     * @param current   The current column state
+     */
+    void columnStateChanged(const DispensingItems& prev, const DispensingItems& current) {
+        char changed[128];
+        *changed = 0;
+        J* body = buildColumnEvent(current);
+        if (current.isOnline() && prev.itemCount != current.itemCount) {
+            strncat(changed, ",items", sizeof(changed)-1);
+        }
+        if (prev.fillState != current.fillState) {
+            strncat(changed, ",state", sizeof(changed)-1);
+        }
+        if (*changed) {
+            JAddStringToObject(body, DATA_FIELD_CHANGES, changed+1);    // skip the initial comma
+        }
+
+        if (prev.isOnline() && current.isOnline()) {
+            // add a text field with human-readable change details
+            const char* text = nullptr;
+            if (prev.fillState != DispensingItems::COLUMN_FULL && current.fillState==DispensingItems::COLUMN_FULL) {
+                text = "Column restocked.";
+            }
+            else if (prev.itemCount < current.itemCount) {
+                text = "Item dispensed.";
+            }
+            if (text) {
+                JAddStringToObject(body, DATA_FIELD_TEXT, text);
+            }
+        }
+        sendNotecardEvent(body, DATA_FILE_VENDING, true);
+    }
+
+public:
+    /**
+     * @brief The next column in the chain. This is used to maintain a linked list of columns.
+     */
+    DispensingColumn* next;
+
+    /**
+     * @brief Construct a new Dispensing Column object.
+     * 
+     * @param i2c           The i2c instance to use for communicating with the ToF sensor.
+     * @param i2cAddress    The i2c address of the ToF sensor.
+     * @param xshut_pin     The `XSHUT` pin for the sensor, which is used to switch the sensor
+     *  on and off.
+     * @param column        The 1-based column ID for this column.
+     */
+    DispensingColumn(TwoWire& i2c, uint8_t i2cAddress, uint8_t xshut_pin, uint8_t column) :
```

Comment:
Prefer one casing strategy. So `i2cAddress` and `xshutPin` or `i2c_address` and `xshut_pin`.